### PR TITLE
Unconfirmed tx cache - CHIA-785

### PR DESCRIPTION
**Pull request by almogdepaz**
### Purpose:

add unconfirmed transaction cache to avoid unnecessary serialisation on each new coin state update 

### Current Behavior:
we pull all unconfirmed transactions from the database on each new coin state update to check which transactions where confirmed

### New Behavior:
use cache to get unconfirmed transactions

### CODE REVIEW

File: chia/_tests/environments/wallet.py [hunk 1]
```diff
@@ -247,9 +247,9 @@ async def change_balances(self, update_dictionary: Dict[Union[int, str], Dict[st
 
     async def wait_for_transactions_to_settle(
         self, full_node_api: FullNodeSimulator, _exclude_from_mempool_check: List[bytes32] = []
-    ) -> List[TransactionRecord]:
+    ) -> List[LightTransactionRecord]:
         # Gather all pending transactions
-        pending_txs: List[TransactionRecord] = await self.wallet_state_manager.tx_store.get_all_unconfirmed()
+        pending_txs: List[LightTransactionRecord] = await self.wallet_state_manager.tx_store.get_all_unconfirmed()
```
Comment (Quexington) [inline chia/_tests/environments/wallet.py:252]: Can we add a separate `get_all_unconfirmed_light` so changes like this aren't necessary?  I'm worried about having two different types causing a lot of headache upstream a la `Program`/`SerializedProgram`.

File: chia/_tests/environments/wallet.py [hunk 2]
```diff
@@ -247,9 +247,9 @@ async def change_balances(self, update_dictionary: Dict[Union[int, str], Dict[st
...
         # Filter clawback txs
         pending_txs = [
```
Comment (arvidn) [inline chia/_tests/environments/wallet.py:254]: It seems to me that asking for *all* unconfirmed here is the root of the issue. Did you explore asking only for specific unconfirmed transactions? e.g. have the database query filter for `CLAWBACK_INCOMING_TRANSACTION_TYPE`

File: chia/_tests/environments/wallet.py [hunk 3]
```diff
@@ -345,7 +345,9 @@ async def process_pending_states(self, state_transitions: List[WalletStateTransi
             try:
                 await self.full_node.check_transactions_confirmed(env.wallet_state_manager, txs)
             except TimeoutError:  # pragma: no cover
-                unconfirmed: List[TransactionRecord] = await env.wallet_state_manager.tx_store.get_all_unconfirmed()
+                unconfirmed: List[LightTransactionRecord] = (
+                    await env.wallet_state_manager.tx_store.get_all_unconfirmed()
+                )
                 raise TimeoutError(
                     f"ENV-{i} TXs not confirmed: {[tx.to_json_dict() for tx in unconfirmed if tx in txs]}"
```
Comment (arvidn) [inline chia/_tests/environments/wallet.py:352]: for example; we ask for all unconfirmed here, but then we immediately filter to only include the ones in `txs`

File: chia/wallet/wallet_transaction_store.py [hunk 4]
```diff
@@ -138,6 +145,8 @@ async def add_transaction_record(self, record: TransactionRecord) -> None:
             await conn.execute_insert(
                 "INSERT OR REPLACE INTO tx_times VALUES (?, ?)", (record.name, bytes(record.valid_times))
             )
+            if record.confirmed is False and record not in self.unconfirmed_txs:
+                self.unconfirmed_txs.append(get_light_transaction_record(record))
```
Comment (arvidn) [inline chia/wallet/wallet_transaction_store.py:149]: is it possible for an attacker to send coin updates that a victim wallet will keep adding to this list? until running out of RAM?

**Review by arvidn — APPROVED** [review]

**Pull request merged**